### ISSUE TITLE
Error when running against a large number of accounts in Organizations

### ISSUE THREAD

**Issue by sam-fakhreddine** [aws-samples/bulk-policy-migrator-scripts-for-account-cost-billing-consoles#5]
When running it against the folder where the affected policies for the entire org are in 1 JSOn we get errors about policy size exceeding 10240 or 6144.

I had to write a helper script to split it into account by account:

```python
import os
import concurrent.futures
import subprocess
import boto3

def run_identify_script(account_id):
    # Run the identify_affected_policies.py script for the given account
    subprocess.run(["python3", "identify_affected_policies.py", "-a", account_id])

def run_update_script(directory):
    # Run the update_affected_policies.py script for the given directory
    subprocess.run(["python3", "update_affected_policies.py", "--affected-policies-directory", directory])

def main():
    # Create an organizations client
    org = boto3.client('organizations')

    # Get the list of AWS accounts
    response = org.list_accounts()
    accounts = [account['Id'] for account in response['Accounts']]
    
    # Print the accounts that will be checked
    print(f"The following accounts will be checked: {', '.join(accounts)}")

    # Use a ThreadPoolExecutor to run the identify_affected_policies.py script for each account, 5 at a time
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
        executor.map(run_identify_script, accounts)

    # Get the list of directories that start with "Affected"
    directories = [dir for dir in os.listdir('.') if dir.startswith('Affected')]

    # Use a ThreadPoolExecutor to run the update_affected_policies.py script for each directory, 5 at a time
    with concurrent.futures.ThreadPoolExecutor(max_workers=5) as executor:
        executor.map(run_update_script, directories)

if __name__ == "__main__":
    main()
```

By doing this I was able to apply with no errors about policy sizes.

**Comment by coverprice** [comment]
I have encountered the same problem. The root cause appears to be a bug in the `identify_affected_policies.py` script: as an example, below is a (pared down and redacted) snippet of my `affected_policies_and_suggestions.json`. The key elements to note are:
- The same policy exists across multiple accounts
- The `SuggestedPolicyStatementsToAppend` section contains duplicate entries, apparently 1 for each account listed in `ImpactedPolicies`.

When attempting to apply the configuration, the updater appears to be trying to apply all the `SuggestedPolicyStatementsToAppend` entries at once, which manifests the bug in 1 of 2 ways:
- The total size of the new policy is so large that the AWS "update policy" API call fails with an error, e.g. "`LimitExceeded: Cannot exceed quota for PolicySize: 6144`"
- The total size is small enough to fit within the size limit, but the repeat sections each use the same SID ("`BillingConsolePolicyMigrator0`"), so the API "update policy" call rejects it with an error like "`MalformedPolicyDocument: Statement IDs (SID) in a single policy must be unique.`"

Running the `identify_affected_policies.py` script on 1 account at a time, as @sam-fakhreddine suggests above, avoids the issue.

```
    {
        "GroupName": "Group7",
        "ImpactedPolicies": [
            {
                "Account": "12345678901",
                "PolicyType": "GroupInlinePolicy",
                "PolicyName": "AWSPortalAccess",
                "PolicyIdentifier": "business"
            },
            {
                "Account": "987654321091",
                "PolicyType": "CustomerManagedPolicy",
                "PolicyName": "BillingFullAccess",
                "PolicyIdentifier": "arn:aws:iam::987654321091:policy/BillingFullAccess"
            }
        ],
        "ImpactedPolicyStatements": [
            {
                "Effect": "Allow",
                "Action": [
                    "aws-portal:*"
                ],
                "Resource": [
                    "*"
                ]
            }
        ],
        "SuggestedPolicyStatementsToAppend": [
            {
                "Action": [
                    "account:CloseAccount",
                    "account:DeleteAlternateContact",
...
[long list of permissions removed for clarity]
...
                    "tax:PutTaxInheritance"
                ],
                "Sid": "BillingConsolePolicyMigrator0",
                "Resource": [
                    "*"
                ],
                "Effect": "Allow"
            },
            {
                "Sid": "BillingConsolePolicyMigrator0",
                "Effect": "Allow",
                "Action": [
                    "account:CloseAccount",
                    "account:DeleteAlternateContact",
...
[long list of permissions removed for clarity]
...
                   "tax:PutTaxInheritance"
                ],
                "Resource": [
                    "*"
                ]
            }
        ]
    },
```

**Comment by coverprice** [comment]
I think I found the source of the bug. 

In the `identify_affected_policies.py` script there is a `merge_maps` function. From what I can tell, it gathers the various policies from individual accounts, and groups them together. The [following snippet from that function](https://github.com/aws-samples/bulk-policy-migrator-scripts-for-account-cost-billing-consoles/blob/main/policy_migration_scripts/scripts/identify_affected_policies.py#L75-L78
) iterates over the suggested replacements. The `k` here appears to be the policy's hash, and the `v` is the policy statement dict.

```python
    for k, v in individual.policy_hash_to_suggested_replacements.items():
        if k not in aggregate.policy_hash_to_suggested_replacements:
            aggregate.policy_hash_to_suggested_replacements[k] = []
        aggregate.policy_hash_to_suggested_replacements[k].extend(v)
```

If my understanding is correct, the above code extends the list of suggested replacements, when instead it should just be setting it. i.e.

```python
    for k, v in individual.policy_hash_to_suggested_replacements.items():
        if k not in aggregate.policy_hash_to_suggested_replacements:
            aggregate.policy_hash_to_suggested_replacements[k] = v
```

Since `k` is a hash that uniquely identifies the policy contents, there is no point in creating a list of the policy values (which are all equivalent); only 1 is needed. 



Incidentally, I found that the [way the hash is generated](https://github.com/aws-samples/bulk-policy-migrator-scripts-for-account-cost-billing-consoles/blob/main/policy_migration_scripts/utils/hashing.py#L73C3-L82) could be improved a little: It dumps the policy to JSON, sorts the resulting string, then hashes the result. However it's not necessary to sort the string; just use the `sort_keys=True` parameter of `json.dumps()`, e.g.

```python
>>> import json
>>> a = {'foo': 1, 'bar': 2}
>>> b = {'bar': 2, 'foo': 1}
>>> json.dumps(a) == json.dumps(b)
False
>>> json.dumps(a, sort_keys=True) == json.dumps(b, sort_keys=True)
True
```
